### PR TITLE
[EC-42] Create cleaner client interface to Skyhook functionality

### PR DESCRIPTION
**Pull request by lzhao-skyhook**
1. clean up v2 protocol interface
2. add easy interfaces for clients

### CODE REVIEW

**Comment by lzhao-skyhook** [comment]
@tboinske-skyhook - please review.

File: inc/sky_protocol.h [hunk 1]
```diff
@@ -646,16 +592,65 @@ int32_t sky_encode_resp_bin(uint8_t *buff, uint32_t buff_len,
...
+ * Skyhook Easy APIs for ELGv2 Protocol client
+ *
+ *************************************************************************/
+
+// parse url out for host and port
+// @param url - in format of "elg://host:port/"
+// @param host - host name in url
+// @param port - port number in url
+// @return true for success or false for failure
+bool sky_parse_url(char * url, char * host, uint16_t * port);
```
Comment (tboinske-skyhook) [inline inc/sky_protocol.h]: It's not clear which parameters are input vs. output. For output parameters, how big do they need to be? E.g., presumably host is an output parameter, meaning that the caller of this function must allocate space for it. How much space should he allocate?

File: inc/sky_protocol.h [hunk 2]
```diff
@@ -646,16 +592,65 @@ int32_t sky_encode_resp_bin(uint8_t *buff, uint32_t buff_len,
...
+// @param len - buffer length
+void print_buff(uint8_t *buff, uint32_t len);
+
+// print binary buffer in HEX to array
+// @param buff - binary buffer
+// @param len - buffer length
+// @param hex_buff - HEX buffer
+// @param hex_buff_len - HEX buffer length
+// @return the data length in HEX buffer for success, or -1 for failure
+int32_t sprint_buff(uint8_t *hex_buff, uint32_t hex_buff_len, uint8_t *buff, uint32_t buff_len);
```
Comment (tboinske-skyhook) [inline inc/sky_protocol.h]: Similar comment - can't tell which parameters are input vs. output.

File: inc/sky_protocol.h [hunk 3]
```diff
@@ -646,16 +592,65 @@ int32_t sky_encode_resp_bin(uint8_t *buff, uint32_t buff_len,
...
+
+// client encodes, encrypts and sends client's location request to Skyhook location service
```
Comment (tboinske-skyhook) [inline inc/sky_protocol.h]: // Called by the client to encode, encrypt and send a location request to the Skyhook location service.

File: inc/sky_protocol.h [hunk 4]
```diff
@@ -646,16 +592,65 @@ int32_t sky_encode_resp_bin(uint8_t *buff, uint32_t buff_len,
...
+int32_t sky_send_location_request(uint8_t *buff, uint32_t buff_len,
+        struct location_rq_t * rq,
+        sky_tx_fn tx, char * url, void * rpc_handler);
+
+// client receives, decrypts and decodes Skyhook location service's response
+// @param buff - data buffer to store the received data
+// @param buff_len - the length of data buffer
+// @param rsp - server's location response
+// @param rx - callback function for receiving data
+// @param rpc_handler - the RPC call handler to mask the underlying communication details
```
Comment (tboinske-skyhook) [inline inc/sky_protocol.h]: Once I have a better understanding of how these functions are used, I'll likely have additional suggestions for improving the comments here.

File: src/protocol/sky_protocol.c [hunk 5]
```diff
@@ -14,6 +15,18 @@
 #include "sky_crypt.h"
 #include "sky_protocol.h"
 
+// parse url
+inline
+bool sky_parse_url(char * url, char * host, uint16_t * port) {
+    uint32_t val = 0;
+    if (sscanf(url, "elg://%[^:]%*[:]%u/", host, &val) != 2) {
+        perror("ERROR: sky_parse_url() received wrong url; expected url format is 'elg://host:port/'");
```
Comment (tboinske-skyhook) [inline src/protocol/sky_protocol.c]: Do you need to test for and handle overflow conditions? What if the buffer pointed to by host is not large enough?

File: inc/sky_protocol.h [hunk 6]
```diff
@@ -531,95 +541,32 @@ struct location_rsp_t {
...
+// callback function for sending data from buffer
+// @param buff - data buffer
+// @param buff_len - data length in buffer
+// @param host - host name
+// @param port - port number
+// @param rpc_handler - the remote procedure call handler (e.g. socket) for client-server model communication;
+//                    - value is set within sky_tx_fn, and is used by sky_rx_fn.
+// @return the number of sent bytes, which should be equivalent to buff_len, upon success,
+//         or -1 upon failure.
+typedef int32_t (* sky_tx_fn)(uint8_t *buff, uint32_t buff_len,
```
Comment (tboinske-skyhook) [inline inc/sky_protocol.h]: Suggest using send instead of tx, recv instead of rx, as these suggest the corresponding Berkeley socket functions.

**Comment by lzhao-skyhook** [comment]
Addressed all code review comments. Also added a simple query client API to consolidate send and recv.
